stellargraph/layer/node2vec.py

```diff
@@ -0,0 +1,195 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+"""
+node2vec
+
+"""
+__all__ = ["Node2Vec"]
+
+from tensorflow.keras import Input
+from tensorflow.keras import backend as K
+from tensorflow.keras.layers import Reshape, Embedding
+import math
+from tensorflow import keras
+import warnings
+
+from ..mapper import Node2VecLinkGenerator, Node2VecNodeGenerator
+
+
+class Node2Vec:
+    """
+    Implementation of the node2vec algorithm of A. Grover and J. Leskovec with Keras layers.
+    see: https://snap.stanford.edu/node2vec/
+
+    The model minimally requires specification of the embedding size and a generator object.
+
+    Args:
+        emb_size (int): The dimension of node embeddings.
+        generator (Sequence): A NodeSequence or LinkSequence.
+        node_num (int): The number of nodes in the given graph.
+    """
+
+    def __init__(self, emb_size, generator=None, **kwargs):
+
+        # Get the node_num from the generator if it is given
+        self.generator = generator
+        if generator is not None:
+            self._get_sizes_from_generator(generator)
+        else:
+            self._get_sizes_from_keywords(kwargs)
+
+        # Model parameters
+        self.emb_size = emb_size
+        self.initializer = (
+            "glorot_uniform"
+        )  # the initializer for the weights to construct hidden layers
+
+    def _get_sizes_from_generator(self, generator):
+        """
+        Sets node_num and multiplicity from the generator.
+        Args:
+             generator: The supplied generator.
+        """
+        if not isinstance(generator, (Node2VecNodeGenerator, Node2VecLinkGenerator)):
+            raise TypeError(
+                "Generator should be an instance of Node2VecNodeGenerator or Node2VecLinkGenerator"
+            )
+
+        self.multiplicity = generator.multiplicity
+        self.input_node_num = generator.graph.number_of_nodes()
+
+        if len(list(generator.graph.node_types)) > 1:
+            raise RuntimeError("Node2Vec called on graph with more than one node type.")
+
+    def _get_sizes_from_keywords(self, kwargs):
+        """
+        Sets node_num and input_feature_size from the keywords.
+        Args:
+             kwargs: The additional keyword arguments.
+        """
+        try:
+            self.input_node_num = kwargs["node_num"]
+            self.multiplicity = kwargs["multiplicity"]
+
+        except KeyError:
+            raise KeyError(
+                "Generator not provided; node_num and multiplicity must be specified."
+            )
+
+    def __call__(self, xin):
+        """
+        Construct node representations from node ids through a look-up table.
+
+        Args:
+            xin (Keras Tensor): Batch input node ids.
+
+        Returns:
+            Output tensor.
+        """
+        input_initializer = keras.initializers.RandomUniform(minval=-1.0, maxval=1.0)
+        input_embedding = Embedding(
```

Comment:
I think layers should be initialized in the `__init__` method, and applied in `__call__` - see, e.g., the GAT class. 
This would allow you to simplify the `link_model` method below to 
```
x_inp_src, x_out_src = self.node_model()
x_inp_dst, x_out_dst = self.node_model()
...
```

Reply:
Yeah, its better to place the initialisation to the __init__ method. But we should make a difference about the definition of (x_inp_src, x_out_src) and (x_inp_dst, x_out_dst), as they use different embedding look-up tables according to the word2vec paper and its original C implementation.

Reply:
Fixed